### PR TITLE
relu ref

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #78601
* #78604
* #78527

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: torch/_refs/nn/functional/__init__.py [hunk 1]
```diff
@@ -127,6 +128,24 @@ def elu(
...
+    a: TensorLikeType, inplace: bool = False
+) -> TensorLikeType:
+    """
+    Reference implementation of torch.nn.functional.relu
+    """
+
+    if inplace:
+        raise NotImplementedError
+
+    return torch.where(torch.gt(a, 0), a, torch.zeros_like(a))
```
Comment (ngimel) [inline torch/_refs/nn/functional/__init__.py:146]:
```suggestion
    return torch.where(torch.gt(a, 0), a, 0)
```
to avoid materializing full size 0 tensor.

Comment (ngimel) [reply torch/_refs/nn/functional/__init__.py]: This doesn't propagate `nans` like eager does:
```
In [1]: import torch

In [2]: a=torch.tensor([float('nan'), -1., 1., 0.])

In [3]: torch.relu(a)
Out[3]: tensor([nan, 0., 1., 0.])

In [4]: torch.where(torch.gt(a, 0), a, 0)
Out[4]: tensor([0., 0., 1., 0.])
```
On one hand, I have been on record saying "activations should use `where` where appropriate, it makes more sense and leads to more sensible gradients where people have no reason to ask why aren't you splitting grad when values are equal, like minimum and maximum do". On another, this breaks with existing tradition of aggressively propagating nans, and we are definitely biasing towards existing nan propagation behavior.

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -19456,6 +19456,14 @@ def __init__(
             DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref_meta',),
         ),
     ),
+    PythonRefInfo(
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:19459]: relu is an elementwise unary operation, so we should make this (and the relu OpInfo) elementwise unary (that should also test the nan propagation)

File: torch/_refs/nn/functional/__init__.py (hunk 1 above)
Comment (ezyang) [reply torch/_refs/nn/functional/__init__.py]: done

**Review by ngimel — APPROVED** [review]
Lint's failing
